### ISSUE TITLE
Non-PHAR installation of Composer not setting environment correctly due to newer symfony/process version

### ISSUE THREAD

**Issue by icedream** [composer/composer#9930]
First of all, I'm testing this on **PHP 7.4.19** and with **Composer 2.0.14**, but from what I figured out I think the software versions are irrelevant for the underlying issue, especially if not installed via the official PHAR file.

### `composer.json`

```json
{
  "scripts": {
    "export": "sh -c export"
  }
}
```

So we have a script that displays the environment variables it received.

### Correct behavior (PHAR installation of Composer)

If we use the **officially released `composer.phar`** for version 2.0.14 (current one at the time of writing this), everything works as normal when I run `php composer.phar run export` (I snipped out the relevant environment variables):

```
export COMPOSER_BINARY='/composer/test-project/composer.phar'
export PATH='/composer/test-project/vendor/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
export PWD='/composer/test-project'
```

 `COMPOSER_BINARY` is set by Composer, `PWD` is where I put this composer.json and `PATH` is expanded by Composer for accessing vendor bins. All as expected so far.

### Incorrect, current behavior (non-PHAR installation of Composer)

Now, I install an in-project version of Composer with `composer require composer/composer:2.0.14`. This should in theory result in the exact same version of Composer being installed, however the difference now will be that dependencies of `composer/composer` are no longer bound to a lockfile and will instead be a newer compatible version. I'm getting back to this further down in this issue description.

However when I now run `vendor/bin/composer run export` the following can be observed:

```
export PATH='/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
```

The `COMPOSER_BINARY` environment variable has gone completely missing and `PATH` has not been expanded. This also means that if I were to try to run a Composer script that itself ran something that is installed to the bin directory, it would no longer be available in PATH and so that script can fail if it doesn't explicitly call it with `vendor/bin/` prefixed.

There is a difference in this output between when PHP is configured with `variables_order = "EGPCS"` (resulting in the output above) and when PHP is configured with `variables_order = "GPCS"` which results in the following:

```
export PATH='/composer/test-project/vendor/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
```

Now `PATH` is correctly expanded, making Composer scripts calling vendor bin commands work again, however `COMPOSER_BINARY` is still missing.

### Source of the problem

The crucial part for calling subprocesses in Composer is [this `Process` instantiation part of `ProcessExecutor`](https://github.com/composer/composer/blob/2.0.14/src/Composer/Util/ProcessExecutor.php#L116-119). `Process` is provided by `symfony/process`, which in the PHAR file – which derives its locked down versions from [Composer's own lockfile](https://github.com/composer/composer/blob/2.0.14/composer.lock#L1047) – is installed as version 2.8.52. In this version, Symfony [defaults to `null` for the environment variables](https://github.com/symfony/process/blob/v2.8.52/Process.php#L137) and [passes that `null` value to `proc_open`](https://github.com/symfony/process/blob/v2.8.52/Process.php#L285). This correctly passes through all environment variables, plus those set by `putenv`.

This behavior [changed in 3.3.15 and a default environment variable set is generated from a combination of `$_SERVER` and `$_ENV`](https://github.com/symfony/process/compare/v3.3.14...v3.3.15#diff-a82f1f0b74256d93193ad7436778b39972f484e9286f70f203ef5b2aab052bc7R1727-R1744), with `$_ENV` taking higher priority. However `$_ENV` is its own set of values that would not be modified by calls to `putenv` as is done by Composer in several places such as [here for `COMPOSER_BINARY`](https://github.com/composer/composer/blob/2.0.14/bin/composer#L60), and it would even override updates in `$_SERVER` such as [here for `PATH`](https://github.com/composer/composer/blob/2.0.14/src/Composer/EventDispatcher/EventDispatcher.php#L543). This means even if Composer updates these values via `putenv` or `$_SERVER`, it would get overwritten for subprocesses because the old value is still in `$_ENV`. `getenv` is not considered at all, so all `putenv` changes simply get lost.

This explains why with `variables_order = "GPCS"` the `PATH` value would correctly be passed through (since it is set via `$_SERVER` in addition to `putenv`), but with `variables_order = "EGPCS"` not. It also explains why `COMPOSER_BINARY` is completely lost since it is only set via `putenv`. In either case though, this is an incompatibility with newer `symfony/process` versions.

I am currently considering opening an issue ticket over at Symfony's repository for this as well, however I am trying to make the developers and users of Composer aware of this issue since it breaks what I consider pretty important functionality in Composer when it is installed as a project's dependency where a PHAR is pretty much out of question.

**Comment by Seldaek** [comment]
You can open an issue with Symfony but I think it's unlikely they will change this at this point.

I am well aware of the issue and that's why all our putenv()'s are accompanied by $_SERVER writes. The COMPOSER_BINARY one missing is an omission which is now fixed in e0366bc76.

@nicolas-grekas as you mostly did this change in https://github.com/symfony/process/commit/d7c95ff7eec4722f85d7b0615ca32960e4c5c66f do you maybe have a hint here how to best handle this?

I am not so keen on adding $_ENV writes on top of the $_SERVER writes as it's just getting very messy, but if it's the only way I guess maybe we should.. I guess this is rarely a problem because the recommended variable order is GPCS both for prod and dev.

**Comment by Seldaek** [comment]
OK had a chat with @nicolas-grekas, I guess I'll add $_ENV sets everywhere. Working on a fix.
